runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,458 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+         }
+      else
+         {
+         generateMemSrc1Instruction(cg, TR::InstOpCode::stw, node, new (cg->trHeapMemory()) TR::MemoryReference(dataSnippetRegister, offsetInDataBlock, 4, cg), resultReg->getLowOrder());
+         generateMemSrc1Instruction(cg, TR::InstOpCode::stw, node, new (cg->trHeapMemory()) TR::MemoryReference(dataSnippetRegister, offsetInDataBlock + 4, 4, cg), resultReg->getHighOrder());
+         }
+      
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, endLabel);
+      }
+   generateReportOOL->swapInstructionListsWithCompilation();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel);
+
+   cg->stopUsingRegister(scratchReg);
+   cg->stopUsingRegister(cndReg);
+   cg->stopUsingRegister(cpIndexReg);
+   cg->stopUsingRegister(resultReg);
+   }
+
+/*
+ * Generate the reporting field access helper call with required arguments
+ *
+ * jitReportInstanceFieldRead
+ * arg1 pointer to static data block
+ * arg2 object being read
+ *
+ * jitReportInstanceFieldWrite
+ * arg1 pointer to static data block
+ * arg2 object being written to
+ * arg3 pointer to value being written
+ *
+ * jitReportStaticFieldRead
+ * arg1 pointer to static data block
+ *
+ * jitReportStaticFieldWrite
+ * arg1 pointer to static data block
+ * arg2 pointer to value being written
+ *
+ */
+void generateReportFieldAccessOutlinedInstructions(TR::Node *node, TR::LabelSymbol *endLabel, TR::Register *dataBlockReg, bool isWrite, TR::RegisterDependencyConditions *deps, TR::CodeGenerator *cg, TR::Register *sideEffectRegister, TR::Register *valueReg)
+   {
+   bool isInstanceField = node->getOpCode().isIndirect();
+   J9Method *owningMethod = reinterpret_cast<J9Method *>(node->getOwningMethod());
+
+   TR_RuntimeHelper helperIndex = isWrite ? (isInstanceField ? TR_jitReportInstanceFieldWrite: TR_jitReportStaticFieldWrite):
+                                            (isInstanceField ? TR_jitReportInstanceFieldRead: TR_jitReportStaticFieldRead);
+
+   TR::Linkage *linkage = cg->getLinkage(runtimeHelperLinkage(helperIndex));
+   auto linkageProperties = linkage->getProperties();
+   TR::Register *valueReferenceReg = NULL;
+
+   TR_BackingStore * location = NULL;
+   TR::Compilation * comp = cg->comp();
+   TR_J9VMBase *fej9 = static_cast<TR_J9VMBase *> (cg->fe());
+
+   /*
+    * For reporting field write, reference to the valueNode is needed so we need to store
+    * the value on to a stack location first and pass the stack location address as an arguement
+    * to the VM helper
+    */
+   if (isWrite)
+      {
+      TR::DataType dt = node->getDataType();
+      int32_t elementSize = dt == TR::Address ? TR::Compiler->om.sizeofReferenceField() : TR::Symbol::convertTypeToSize(dt);
+      TR::InstOpCode::Mnemonic storeOp = getLoadOrStoreFromDataType(cg, dt, elementSize, node->getOpCode().isUnsigned(), false);
+      location = cg->allocateSpill(node->getSize(), false, NULL);
```

Comment:
Can we use `allocateLocalTemp` here? It seems all we care about is the symbol reference. The advantage of using `allocateLocalTemp` is you don't have to worry about freeing the spill location at all.

Reply:
I'm not sure what affects this would have. In theory it should work, but due to the time strain I would rather create an issue to replace this at a later time where we can thoroughly test this change.

Reply:
It has the effect of not having to call `freeSpill` at the end of the function. I disagree w.r.t. not making code improvements due to time constraints. We should not be in the habit of sacrificing quality because of arbitrary deadlines as technical debt will consume us if we continue down this path. This is a rather simple improvement and perhaps the best time to make it a reality.

Reply:
@fjeremic we did not plan to leave it as is but open another PR. I am just not sure how much time we have left for 0.15 @pshipton

Reply:
> @fjeremic we did not plan to leave it as is but open another PR. I am just not sure how much time we have left for 0.15 @pshipton

Given we've deferred twice already I'm not sure what the rush here is as there doesn't seem to user demand for this feature on Power (as far as I'm aware). Regardless, this is a simple three line change which does not warrant a whole PR. We are already reviewing the entire change here so why delay making the improvement? Let's just do it now. We've spent more effort typing responses on this thread than it would take to make the change.

Reply:
See https://github.com/eclipse/openj9/pull/6234/commits/59478ec9e611a0f51d70efd62ff92055afca576a